include/dd/Verification.hpp

```diff
@@ -0,0 +1,66 @@
+#include "dd/FunctionalityConstruction.hpp"
+#include "dd/Package.hpp"
+
+namespace dd {
+
+// get next garbage qubit before n
+inline Qubit getNextGarbage(Qubit n, const std::vector<bool>& garbage) {
+  while (n > 0 && !garbage.at(n)) {
+    --n;
+  }
+  return n;
+}
+/**
+    Checks for partial equivalence between the two circuits c1 and c2.
+    Assumption: the data qubits are all at the end of the input qubits and
+    the input and output permutations are the same.
+
+    @param circuit1 First circuit
+    @param circuit2 Second circuit
+    @return true if the two circuits c1 and c2 are partially equivalent.
+    **/
+template <class Config>
+bool partialEquivalenceCheck(qc::QuantumComputation c1,
+                             qc::QuantumComputation c2,
+                             std::unique_ptr<dd::Package<Config>>& dd) {
+
+  auto d1 = c1.getNqubitsWithoutAncillae();
+  auto d2 = c2.getNqubitsWithoutAncillae();
+  auto m1 = c1.getNqubits() -
+            static_cast<std::size_t>(std::count(c1.getGarbage().begin(),
+                                                c1.getGarbage().end(), true));
+  auto m2 = c2.getNqubits() -
+            static_cast<std::size_t>(std::count(c2.getGarbage().begin(),
+                                                c2.getGarbage().end(), true));
+  if (m1 != m2 || d1 != d2) {
+    return false;
+  }
+
+  // add swaps in order to put the measured (= not garbage) qubits in the end
+  auto garbage = c1.getGarbage();
+  auto n = static_cast<Qubit>(garbage.size());
+  auto nextGarbage = getNextGarbage(n - 1, garbage);
+  // find the first garbage qubit at the end
+  for (Qubit i = 0U; i < n - static_cast<Qubit>(m1); i++) {
+    if (!garbage.at(i)) {
+      // swap it to the end
+      c1.swap(i, nextGarbage);
+      c2.swap(i, nextGarbage);
+      --nextGarbage;
+      nextGarbage = getNextGarbage(nextGarbage, garbage);
+    }
+  }
```

Comment:
For now I just assumed that these two preconditions are already satisfied (and I also wrote it as a comment over the function), because I thought that some code from [mqt-qcec](https://github.com/cda-tum/mqt-qcec) can be reused to enforce these preconditions.

Reply:
I guess it is fine to assume the pre-conditions for now.
As stated above, the first one is kind of handled naturally in mqt-core as well as mqt-qcec via the `initialLayout`.
The latter pre-condition is not enforced at all, but i think it should hold for most practical cases (where the first `n` logical qubits are the data qubits).
If these pre-confitions are assumed anyway, then I guess it is fair to remove the above code snippet.